Pass autocommit context to connector on starting a transaction

hantangwangd opened the pull request:
## Description

Currently, we only have a connector-level method `Connector.isSingleStatementWritesOnly` to determine whether the entire connector supports single-statement writes exclusively, that is, whether it only permits execution within auto commit transactions.

However, some connectors have the capability not to be single-statement writes only, but certain statements (such as some non-rollbackable DDL statements) should not be placed in non-autocommit transactions. Therefore, the transaction's autocommit context needs to be passed to the connectors, allowing them to determine whether individual statements are permitted in non-autocommit transactions.

## Motivation and Context
<!---Why is this change required? What problem does it solve?-->
<!---If it fixes an open issue, please link to the issue here.-->

## Impact
<!---Describe any public API or user-facing feature change or any performance impact-->

## Test Plan
<!---Please fill in how you tested your change-->

## Contributor checklist

- [ ] Please make sure your submission complies with our [contributing guide](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md), in particular [code style](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#code-style) and [commit standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes

```
== NO RELEASE NOTE ==
```

tdcmeehan commented:
I am a bit confused about the intent of this PR.  Could you help to provide some examples on a scenario in Iceberg where it is helpful to add this flag?  I note that the `isSingleStatementWritesOnly` flag is still true in this PR for the Iceberg connector.

I also don't see where the `autoCommitContext` is actually used in this PR.  What changes once this PR is merged?

hantangwangd commented:
@Tdcmeehan, thanks for the review and comment. I think the core purpose of this PR is to support connectors that allow multi-write transaction for a subset of SQL statements. Currently, for many connectors (like Iceberg), multi-write statement transaction support is typically limited to some certain DML operations, with little to no support for DDL statements. It's not an all-or-nothing proposition, so a connector-level `isSingleStatementWritesOnly` flag appears insufficient for this purpose. We need a way to let each connector determine - based on the current transaction's `autoCommitContext` flag - whether a given statement should be permitted execution within the current transaction.

Take Iceberg for example, with this `autoCommitContext` flag passed in, it can mark it's connector level flag `isSingleStatementWritesOnly` as false, and explicitly check and throw an exception when execute DDL statements like `DROP TABLE` in a transaction with flag `autoCommitContext=false`.

> I also don't see where the autoCommitContext is actually used in this PR. What changes once this PR is merged?

This PR do not really open the multi-write statement transaction for Iceberg. After this is merged, I plan to support the actual multi-statements write transaction for Iceberg in one or several follow-up PRs. What's your opinion?

tdcmeehan commented:
@hantangwangd can you include the Iceberg changes into this PR as a separate commit?

hantangwangd commented:
> can you include the Iceberg changes into this PR as a separate commit?

Sure, I'll add the change later. My thought is, in the beginning, we can allow DQL and some certain DML to run together within a transaction with `autoCommitContext=false`, under the snapshot isolation. Then, we can progressively expand the range of statements permitted in multi-write statement transactions based on the specific characteristics of each Iceberg operation. Do you think this makes sense? @tdcmeehan

hantangwangd commented:
Hi @tdcmeehan, I've added the code changes to support multi-statement writes transactions for Iceberg, and added the relevant design choices to the PR description. Please take a look when convenient. Any feedback would be greatly appreciated!

hantangwangd commented:
@steveburnett thank you for the reminder. Done! Please take a look when you get a chance.

tdcmeehan reviewed (commented):
Looks really great.  I only have some nits.
